src/vos/vos_space.c

```diff
@@ -303,19 +303,42 @@ vos_space_hold(struct vos_pool *pool, uint64_t flags, daos_key_t *dkey,
 	if (scm_left < space_est[DAOS_MEDIA_SCM])
 		goto error;
 
-	/* If NVMe isn't configured or this update doesn't use NVMe space */
-	if (pool->vp_vea_info == NULL || space_est[DAOS_MEDIA_NVME] == 0)
-		goto success;
+	/* If NVMe is configured and this update uses NVMe space */
+	if (pool->vp_vea_info != NULL && space_est[DAOS_MEDIA_NVME] != 0) {
+		nvme_left = NVME_FREE(&vps);
 
-	nvme_left = NVME_FREE(&vps);
-	if (nvme_left < NVME_SYS(&vps))
-		goto error;
+		if (nvme_left < NVME_SYS(&vps))
+			goto error;
 
-	nvme_left -= NVME_SYS(&vps);
-	/* 'NVMe held' has already been excluded from 'NVMe free' */
+		nvme_left -= NVME_SYS(&vps);
+		/* 'NVMe held' has already been excluded from 'NVMe free' */
 
-	if (nvme_left < space_est[DAOS_MEDIA_NVME])
-		goto error;
+		if (nvme_left < space_est[DAOS_MEDIA_NVME])
+			goto error;
+	}
+
+	/* Check space reserve for rebuild */
+	if (!(flags & VOS_OF_REBUILD) && pool->vp_space_rb != 0) {
+		rb_reserve = SCM_TOTAL(&vps) * pool->vp_space_rb / 100;
+
+		if (SCM_FREE(&vps) < (rb_reserve + POOL_SCM_HELD(pool) +
+				      space_est[DAOS_MEDIA_SCM])) {
+			D_ERROR("Insufficient rebuild reserved SCM space for Pool:"DF_UUID".\n",
+				DP_UUID(pool->vp_id));
+			goto error;
+		}
+
+		if (pool->vp_vea_info == NULL || space_est[DAOS_MEDIA_NVME] == 0)
+			goto success;
+
+		rb_reserve = NVME_TOTAL(&vps) * pool->vp_space_rb / 100;
```

Comment:
Do you think we need different ratio for nvme and scm?

Reply:
Looks not necessary to me, please see my commit comment, we actually even can't honor NVMe reserve in some cases, but I don't know how to improve it so far.